Add necessary changes for supporting the Zephyr DAI SAI driver on i.MX93

LaurentiuM1234 opened the pull request:
Tested locally using SOF with Zephyr on i.MX93.

mmahadevan108 commented:
@mcuxsusan , can you please help assign reviewers to this PR.

mcuxsusan commented:
@mmahadevan108, I have worked internally assigning the issue to our developer, no progress yet. cc @zejiang0jason to help check.

LaurentiuM1234 commented on drivers/sai/fsl_sai.c:
```diff
@@ -1439,12 +1437,14 @@ void SAI_TxSetConfig(I2S_Type *base, sai_transceiver_t *config)
     }
+#endif
 
-    for (i = 0U; i < (uint32_t)FSL_FEATURE_SAI_CHANNEL_COUNTn(base); i++)
```
The issue should be solved now.

mmahadevan108 replied:
@ zejiang0jason, could you kindly revisit this PR.

zejiang0jason commented on drivers/sai/fsl_sai.c:
```diff
@@ -269,7 +269,13 @@ static uint32_t SAI_GetInstance(I2S_Type *base)
+    }
+#else
     assert(instance < ARRAY_SIZE(s_saiBases));
+#endif /* NDEBUG */
```
Hi @LaurentiuM1234 , for the warning, suggest update the SAI_Init and SAI_Deinit, check the return value of SAI_GetInstance, if the value is wrong, then return kStatus_InvalidArgument to upper layer.
If we simply return 0, user will not notice the parameter error.
Thanks.

LaurentiuM1234 replied:
Well we already have an `assert` statement inside `SAI_GetInstance()` that's supposed to make sure the returned instance is valid. The issue appears when asserts are disabled, which not only would result in the compilation warnings I've mentioned but it would break a lot of the other functions as well (in case of bad arguments for instance) (since most functions use asserts to make sure the arguments are valid). If we were to return `kStatus_InvalidArgument` inside `SAI_GetInstance()` that would mean getting rid of the assert (which would become pointless) but the problem is that I don't think it would aid the debugging much anyways as most of the SAI functions return a `void` so there's no way to propagate the errors to upper layers. Of course, if you still think this is a good idea then I'm all up for it, but I guess the core issue here is that the SAI functions rely on asserts for error handling which may not always be enabled.

zejiang0jason replied:
thank you, I agree with you that to catch this simple invalid base address issue, assert is easier than use function return code.
how about like this?
```c
if (instance == ARRAY_SIZE(s_saiBases))
{
    assert(false);
    instance = 0;
}
```
thanks.

LaurentiuM1234 replied:
Sounds good to me. Will try it out locally and update the PR if everything is ok.

mcuxted commented on drivers/sai/fsl_sai.c:
```diff
@@ -1228,8 +1228,7 @@ void SAI_TxSetFifoConfig(I2S_Type *base, sai_fifo_t *config)
 {
     assert(config != NULL);
 #if defined(FSL_FEATURE_SAI_HAS_FIFO) && (FSL_FEATURE_SAI_HAS_FIFO)
-    if ((config->fifoWatermark == 0U) ||
```
Hi @LaurentiuM1234 , many old examples may not have watermark configured at the beginning, it will  cause the watermark to be automatically configured as 0. Please add the default configuration config->fifoWatermark = (uint8_t)((uint32_t)FSL_FEATURE_SAI_FIFO_COUNTn(base) / 2U); in SAI_GetCommonConfig.

mcuxted commented on drivers/sai/fsl_sai.c:
```diff
@@ -1419,24 +1419,15 @@ void SAI_RxSetSerialDataConfig(I2S_Type *base, sai_serial_data_t *config)
- * param config transmitter configurations.
- */
-void SAI_TxSetConfig(I2S_Type *base, sai_transceiver_t *config)
+#if !defined(FSL_FEATURE_SAI_DISABLE_IMPLICIT_CHAN_CONFIG) || !FSL_FEATURE_SAI_DISABLE_IMPLICIT_CHAN_CONFIG
```
Hi @LaurentiuM1234, The macro FSL_FEATURE_SAI_xxxxx is the hardware description format. For software macros, please use the MCUX_SDK_SAI_xxxxx format. 

Please also place the macro in fsl_sai.h and add comments.

The method is as follows
/*! @brief xxxxxxx */
#ifndef MCUX_SDK_SAI_xxxxx
#define MCUX_SDK_SAI_xxxxx 0U /* xxxxxxxx */
#endif

LaurentiuM1234 commented:
@mcuxted Thanks for the suggestions! I've updated the PR accordingly. Could you please check again?

As for the TX/RX FIFO watermark issue: I've introduced `MCUX_SDK_SAI_ALLOW_NULL_FIFO_WATERMARK` which will maintain the old behavior if set to 0 (which is the default value).

LaurentiuM1234 commented:
@zejiang0jason I've updated the PR according to your suggestion. Also tested it locally and works fine. Please have another look. Thanks!

zejiang0jason commented:
thanks. @LaurentiuM1234 , I have no other comment, let's wait for @mcuxted .

mcuxted commented:
thanks. @LaurentiuM1234 , I have no other comment too,

LaurentiuM1234 commented:
Great!  @mcuxsusan I believe this is ready for merge
